### PR TITLE
fix cookie handling with different cookie params

### PR DESCRIPTION
**Pull request by tflori**
We need to remove the existing cookie. If the cookie domain, path or something else got altered via `session_set_cookie_params()` then it does not get deleted without these params.

Another thing is the missing refresh of session cookie when you have set a specific lifetime of a session cookie.

### CODE REVIEW

**Comment by duncan3dc** [comment]
@tflori Thanks for contributing :+1: 

Could you add a test that demonstrates the issue, and then we can ensure this change fixes it, and we're protected from future regressions?

**Comment by tflori** [comment]
@duncan3dc that's a problem. it was not tested before and `headers_list()` does not return the cookie that got set in cli mode - I think the hole functionality of `set_cookie()` is disabled in cli.

**Comment by duncan3dc** [comment]
@tflori Ok, I'll take a look when I get a minute and see if I can find a way to test, thanks

**Comment by tflori** [comment]
2 weeks ago.. I like this lib, it's small and easy but I don't want to work with dev branch. So please consider to merge.

**Comment by duncan3dc** [comment]
Hi @tflori, sorry I didn't get chance to resolve this yet, I'll merge it as soon as I can

**Comment by duncan3dc** [comment]
@tflori I've added some tests that use the built in webserver to test [cookies](https://github.com/duncan3dc/sessions/blob/796198b8766147293210a94388d49c9735f24f08/tests/WebTest.php#L67) now, so if you could write a failing test for your issue then we can demonstrate that this pr fixes it, thanks

**Comment by tflori** [comment]
@duncan3dc I've added the tests. In this tests I found out that destroy a session creates a new session with wrong name and sends a second cookie `PHPSESSID`... I fixed this too by calling `$this->init()` instead of `@session_start()`

**Comment by tflori** [comment]
@duncan3dc what about this PR?

**Comment by duncan3dc** [comment]
@tflori It looks like the tests are failing on PHP 5.6, could you make the regular expression less restrictive? Or maybe use the guzzle cookie parsing so that you can just check the relevant parts of the cookie?

**Comment by tflori** [comment]
@duncan3dc i've fixed them already:
https://travis-ci.org/duncan3dc/sessions/builds/183076797

**Comment by duncan3dc** [comment]
@tflori Sorry I wasn't clear, that test still relies on very specific formatting of the cookie, and tests parts of it we don't care about, for instance we aren't testing that it's httponly, so the test shouldn't fail if the cookie returns `"HTTPONLY"`

**Comment by tflori** [comment]
in SessionInstance in line 64 and 189 we using `$cookieParam['httponly']` - we define this param in tests/web/sub/bootstrap.php and this has to be tested too. The cookie we update or delete should exactly match with the cookie that php generates. That was the reason to check if the line also includes `[Hh]ttp[Oo]nly`

**Comment by tflori** [comment]
@duncan3dc I'm waiting for a release with this fix since 11th december. Three weeks ago you had a question. I answered. Everything is ready.

Without progress I will have to fork this package.

**Comment by duncan3dc** [comment]
Sorry for the delay @tflori, but I need to be careful what code I accept in the project, as I then must maintain that code and support users with it indefinitely.

Are you happy for me to merge this and then adjust the tests to only test the things we need to test?

**Comment by duncan3dc** [comment]
Hi @tflori, I've pushed some changes to master that should handle everything you need, you can inject a `Cookie` instance to manage the cookie settings now. Can you check and let me know if there's anything missing/broken from your use case? Sorry again for the delay on this

**Comment by tflori** [comment]
It is not solving anything for non-session cookies. It just adds another class to hold information that is already handles by php itself.

Before I liked your session library - it was small, simple and allows to have simultaneous request using the same session even with session_handler files.

**Comment by duncan3dc** [comment]
Sorry @tflori, the changes I've made mean that the library now respects session cookie parameters and ensures they are used when starting the session. I understood that was the issue.

Please could you help me to understand any remaining issues?

**Comment by tflori** [comment]
`session_set_cookie_params()` can be executed before. Or the Params can be set in the ini file. It worked before too.

There are two remaining issues:

1. PHP does never delete the cookie. If you have a lot of traffic on your page and mostly the people are not signed in and don't have a session - you don't even have to look for a session file (or key in redis) and this makes it some bit faster. But you need to delete the cookie - it will be nice when the session manager handles this.
2. PHP does not refresh the cookie. A lot of websites (including our big friends google and github) store sessions for longer than a browser session - you are still logged in, right? You will have to refresh the cookie to stay logged in. Imagine your cookie is valid for 7 days and it never gets refreshed: after 7 days while browsing on this website you are suddenly logged out (intruder alert - someone must have stolen my cookie). It will be nice if session manager handles this too.

That are my changes and this I'm testing. To make this work the cookie has to be exactly what php is sending including HttpOnly and the one thing I still can't test: secure only.

File: src/SessionInstance.php [hunk 1]
```diff
@@ -188,13 +198,17 @@ public function destroy()
     {
         $this->init();
 
-        # Start the session up, but ignore the error about headers already being sent
-        @session_start();
```
Comment (tflori) [inline src/SessionInstance.php]: this is already done in `::init()`

File: src/SessionInstance.php [hunk 2]
```diff
@@ -188,13 +198,17 @@ public function destroy()
...
+        # Remove the session cookie
+        setcookie(
+            $this->name, "", 1, $this->cookie->getPath(), $this->cookie->getDomain(),
+            $this->cookie->isSecure(), $this->cookie->isHttpOnly()
+        );
 
-        # Clear the session data
-        $_SESSION = [];
```
Comment (tflori) [inline src/SessionInstance.php]: for this the function `session_unset()` exist

File: src/SessionInstance.php (hunk 1 above)
Comment (duncan3dc) [inline src/SessionInstance.php]: It is, but then `session_write_close()` is called meaning emptying `$_SESSION` has no effect, we must restart the session to empty it

File: src/SessionInstance.php (hunk 2 above)
Comment (duncan3dc) [inline src/SessionInstance.php]: The PHP docs recommend doing it this way and not using `session_unset()`

File: src/SessionInstance.php (hunk 1 above)
Comment (tflori) [inline src/SessionInstance.php]: all tests still work. if there is a problem the tests are not showing this.

File: src/SessionInstance.php (hunk 2 above)
Comment (tflori) [inline src/SessionInstance.php]: Where they recommend it? I can't find. Do you have a link?
http://php.net/manual/en/function.session-unset.php

Comment (tflori) [inline src/SessionInstance.php]: Oh I see

Comment (duncan3dc) [inline src/SessionInstance.php]: _"Only use session_unset() for older deprecated code that does not use $_SESSION"_
http://php.net/manual/en/function.session-destroy.php

File: src/SessionInstance.php (hunk 1 above)
Comment (duncan3dc) [inline src/SessionInstance.php]: If you checkout from master, and remove this line, the tests will fail

Comment (tflori) [inline src/SessionInstance.php]: @duncan3dc can't confirm this:
```
[iras@miracle ~/work/sessions]$ git fetch --all
Fetching origin
Fetching upstream
[iras@miracle ~/work/sessions]$ git merge upstream/master
Already up-to-date.
[iras@miracle ~/work/sessions]$ vendor/bin/phpunit --stderr
PHPUnit 5.7.9 by Sebastian Bergmann and contributors.

................................................................. 65 / 77 ( 84%)
............                                                      77 / 77 (100%)

Time: 101 ms, Memory: 6.00MB

OK (77 tests, 120 assertions)
```

Comment (duncan3dc) [inline src/SessionInstance.php]: Is that from your branch? Where you've made significant changes? Try:
```sh
git checkout master
git pull upstream/master
vendor/bin/phpunit --stderr
```

Comment (tflori) [inline src/SessionInstance.php]: yes, the recommended way is to destroy the session and remove the cooke - as it is in my branch. without deleting the cookie session_destroy seems to be useless. I will write another test to see if session_start and session_destroy is required or not.

**Comment by tflori** [comment]
@duncan3dc I added two tests. one to show session_start is required and one to show session_destroy is required. I hope we can come to an end now.

**Comment by duncan3dc** [comment]
Hi Thomas, I've merged your work into master now, could you check it out to see if it fixes your issues?
Thanks again for your patience, I think you've helped to make the library much better and I'm glad you contributed :+1:

**Comment by tflori** [comment]
@duncan3dc yes it works for me but there is untested code. you can remove the cookie params in destroy without any error. also the code will not work for someone that is not using cookies. 

anyway it needs some cleanup now. I don't understand why you not just merged the pr instead of squashed commited something different. 

I'm very unhappy to contribute to this library anymore.

**Pull request closed without merging**